docs/en/lua-developer-faq.md

```diff
@@ -64,108 +88,143 @@ node.restart(); for i = 1, 20 do print("not quite yet -- ",i); end
+| tmr        | `register([id,] interval, mode, function())` |
+| node       | `task.post([task_priority], function)`, `output(function(str), serial_debug)` | 
+| wifi       | `startsmart(chan, function())`, `sta.getap(function(table))` |
+| net.server | `sk:listen(port,[ip],function(socket))`  |
 | net        | `sk:on(event, function(socket, [, data]))`, `sk:send(string, function(sent))`, `sk:dns(domain, function(socket,ip))` |
-| gpio       | `trig(pin, type, function(level))`                                         |
-| mqtt       | `client:m:on(event, function(conn[, topic, data])`                         |
-| uart       | `uart.on(event, cnt, [function(data)], [run_input])`                       |
+| gpio       | `trig(pin, type, function(level))`  |
+| mqtt       | `client:m:on(event, function(conn[, topic, data])`   |
+| uart       | `uart.on(event, cnt, [function(data)], [run_input])`  |
+
+For a comprehensive list refer to the Module documentation on this site.
+
+### So what are the different ways of declaring variables and how is NodeMCU different here?
+
+The following is all standard Lua and is explained in detail in PiL etc., but it is worth summarising here because understanding this is of particular importance in the NodeMCU environment. 
 
+*  All variables in Lua can be classed as globals, locals or upvalues.  But by default any variable that is referenced and not previously declared as `local` is **global** and this variable will persist in the global table until it is explicitly deleted. If you want to see what global variables are in scope then try
+```Lua
+   for k,v in pairs(_G) do print(k,v) end
+```
+*  Local variables are 'lexically scoped', and you may declare any variables as local within nested blocks or functions without affecting the enclosing scope.
+*  Because locals are lexically scoped you can also refer to local variables in an outer scope and these are still accessible within the inner scope. Such variables are know as **upvalues**..
+*  Lua variable can be assigned two broad types of data: **values** such as numbers, booleans, and strings and **references** such as functions, tables and userdata.  You can see the difference here  when you assign the contents of a variable `a` to `b`.  In the case of a value then it is simply copied into `b`.  In the case of a reference, both `a` and `b` now refer to the *same object*, and no copying of content takes place.  This process of referencing can have some counter-intuitive consequences. For example, in the following code by the time it exists, the variable `timer2func` is out of scope.  However a reference to the function has now been stored in the Lua registry by the alarm API call, so it and any upvalues that it uses will persist until it is eventually entirely dereferenced (e.g. by `tmr2:unregister()`.
+```Lua 
+  do
+    local tmr2unc = function() ds.convert_T(true) tmr1:start() end
```

Comment:
local tmr2unc = function() ds.convert_T(true) tmr1:start() end
->
local tmr2func = function() ds.convert_T(true); tmr1:start() end
+ correction of `tmr2unc` on the following line